File: reverse-polish-notation/java/ReversePolishNotationCalculator.java
import java.util.Stack;

public class ReversePolishNotationCalculator {

    public Double evaluate(String expression) {
        Double result = Double.NaN;
        Stack<Double> operand = new Stack<>();

        for (String op : expression.split("\\s")) {
            char maybeOperator = op.charAt(0);
            switch (maybeOperator) {
                case '+':
                case '-':
                case '/':
                case '÷':
                case '*':
                case '×':
                    Double b = operand.pop();
                    Double a = operand.pop();
                    result = evaluateExp(a, b, maybeOperator);
                    operand.push(result);
                    break;
                default:
                    operand.push(Double.valueOf(op));
            }
        }

        return result;
    }

    private Double evaluateExp(Double a, Double b, char operator) {
        Double result = Double.NaN;
        switch(operator) {
            case '+':
                result = a + b;
                break;
            case '-':
                result = a - b;
                break;
            case '/':
            case '÷':
                result = a / b;
                break;
            case '*':
            case '×':
                result = a * b;
                break;
            default:
                result = Double.NaN;
        }
        return result;
    }

}

File: algorithms/reverse-polish-notation/scala/ReversePolishNotationCalculator.scala
import scala.collection.immutable.List

class ReversePolishNotationCalculator {

    def evaluate(expression: String): Double = {

        def sum(a: Double, b: Double) = a + b
        def subtract(a: Double, b: Double) = a - b
        def divide(a: Double, b: Double) = a / b
        def multiply(a: Double, b: Double) = a * b

        def compute(f: (Double, Double) => Double, stack: List[Double]): List[Double] = {
            val b = stack.head;
            val a = stack.tail.head;
            val result = f(a, b)
            result :: stack.tail.tail
        }

        def evaluateExp(expr: String, stack: List[Double]): List[Double] = {
            expr match {
                case "+" => compute(sum, stack)
                case "-" => compute(subtract, stack)
                case "/" | "÷" => compute(divide, stack)
                case "*" | "×" => compute(multiply, stack)
                case _ => expr.toDouble :: stack
            }
        }

        expression
        .split(" ")
        .foldLeft(List[Double]())((stack, expr) => evaluateExp(expr, stack))
        .head

    }

}
